File: app/src/main/java/co/edu/unipiloto/myapplication/db/DBHelper.java
package co.edu.unipiloto.myapplication.db;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

/**
 * Clase auxiliar para la gestión de la base de datos SQLite de la aplicación.
 * Se encarga de la creación, actualización y gestión de versiones de la base de datos.
 */
public class DBHelper extends SQLiteOpenHelper {
    /**
     * Nombre del archivo de la base de datos.
     */
    public static final String DB_NAME = "logistica.db";
    /**
     * Versión de la base de datos. Si se cambia este número, se activará el método onUpgrade.
     * Versión 8: Fuerza la recreación total con un esquema consistente. Requiere desinstalar la app en el dispositivo de desarrollo.
     */
    public static final int DB_VERSION = 8;

    /**
     * Create a DBHelper configured to manage the application's SQLite database.
     *
     * @param context the application context used to access or create the database
     */
    public DBHelper(Context context) {
        super(context, DB_NAME, null, DB_VERSION);
    }

    /**
     * Creates the initial database schema (users, recolectores, direcciones, guia, solicitudes) and their indexes in the provided database.
     *
     * This method is intended to be invoked from onCreate when the database is created for the first time.
     *
     * @param db the writable SQLiteDatabase where tables and indexes will be created
     */
    private void createTables(SQLiteDatabase db) {

        // ==========================================================
        // 1. Tabla de Usuarios (Clientes)
        // Almacena la información de los clientes que realizan envíos.
        // ==========================================================
        db.execSQL("CREATE TABLE users (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "email TEXT UNIQUE NOT NULL," +
                "password_hash TEXT NOT NULL," +
                "full_name TEXT," +
                "phone_number TEXT" +
                ")");

        // ==========================================================
        // 2. Tabla de Recolectores (Personal de Logística)
        // Almacena información sobre el personal interno como conductores y gestores.
        // ==========================================================
        db.execSQL("CREATE TABLE recolectores (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "username TEXT UNIQUE NOT NULL," +
                "password_hash TEXT NOT NULL," +
                "role TEXT NOT NULL CHECK(role IN ('CONDUCTOR', 'GESTOR', 'FUNCIONARIO', 'ANALISTA'))," +
                "zona TEXT," +
                "is_active INTEGER DEFAULT 1" +
                ")");

        // ==========================================================
        // 3. Tabla de Direcciones
        // Almacena las direcciones de recogida y destino de los envíos.
        // ==========================================================
        db.execSQL("CREATE TABLE direcciones (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "user_id INTEGER," +
                "ciudad TEXT NOT NULL," +
                "full_address TEXT NOT NULL," +
                "latitude REAL," +
                "longitude REAL," +
                "created_at INTEGER NOT NULL," +
                "FOREIGN KEY(user_id) REFERENCES users(id)" +
                ")");

        // ==========================================================
        // 4. Tabla de Guía / Tracking
        // Contiene la información de seguimiento y detalles del paquete.
        // ==========================================================
        db.execSQL("CREATE TABLE guia (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "tracking_number TEXT UNIQUE NOT NULL," +
                "descripcion TEXT," +
                "valor REAL DEFAULT 0.0," +
                "peso REAL DEFAULT 0.0" +
                ")");

        // ==========================================================
        // 5. Tabla de Solicitudes (Tabla central)
        // Es la tabla principal que une toda la información de un envío.
        // ==========================================================
        db.execSQL("CREATE TABLE solicitudes (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "user_id INTEGER NOT NULL," +
                "recolector_id INTEGER," +
                "direccion_id INTEGER NOT NULL," +
                "fecha TEXT NOT NULL," +
                "franja TEXT NOT NULL," +
                "notas TEXT," +
                "zona TEXT NOT NULL," +
                "guia_id INTEGER," +
                "estado TEXT NOT NULL DEFAULT 'PENDIENTE'," +
                "confirmation_code TEXT," +
                "created_at INTEGER NOT NULL," +
                "FOREIGN KEY(user_id) REFERENCES users(id)," +
                "FOREIGN KEY(recolector_id) REFERENCES recolectores(id)," +
                "FOREIGN KEY(direccion_id) REFERENCES direcciones(id)," +
                "FOREIGN KEY(guia_id) REFERENCES guia(id))");

        // ÍNDICES para optimizar las búsquedas
        db.execSQL("CREATE INDEX idx_sol_recolector ON solicitudes(recolector_id)");
        db.execSQL("CREATE INDEX idx_guia_tracking ON guia(tracking_number)");
    }

    /**
     * Create the initial database schema when the database is created for the first time.
     *
     * @param db the writable SQLiteDatabase used to execute schema creation statements
     */
    @Override
    public void onCreate(SQLiteDatabase db) {
        // Llama al método para crear todas las tablas
        createTables(db);
    }

    /**
     * Performs a destructive schema migration when the database version increases.
     *
     * <p>If oldV is less than newV, all application tables are dropped and the schema is recreated by
     * calling onCreate(db). This operation permanently deletes existing data.
     *
     * @param db the SQLiteDatabase instance being upgraded
     * @param oldV the current (old) database version
     * @param newV the target (new) database version
     */
    @Override
    public void onUpgrade(SQLiteDatabase db, int oldV, int newV) {
        if (oldV < newV) {
            db.execSQL("DROP TABLE IF EXISTS solicitudes");
            db.execSQL("DROP TABLE IF EXISTS direcciones");
            db.execSQL("DROP TABLE IF EXISTS guia");
            db.execSQL("DROP TABLE IF EXISTS recolectores");
            db.execSQL("DROP TABLE IF EXISTS users");

            onCreate(db);
        }
    }
}

File: app/src/main/java/co/edu/unipiloto/myapplication/db/DBHelper.kt
package co.edu.unipiloto.myapplication.db

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import co.edu.unipiloto.myapplication.security.PasswordHasher // 🏆 Importación de la nueva clase

/**
 * Clase auxiliar para la gestión de la base de datos SQLite de la aplicación.
 * Se encarga de la creación, actualización y gestión de versiones de la base de datos.
 */
class DBHelper(context: Context) : SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {

    companion object {
        /** Nombre del archivo de la base de datos. */
        const val DB_NAME = "logistica.db"

        /**
         * Versión de la base de datos. Incrementada a 10 para forzar la actualización del esquema.
         */
        const val DB_VERSION = 10

        // Nombres de Tablas
        const val TABLE_USERS = "users"
        const val TABLE_RECOLECTORES = "recolectores"
        const val TABLE_DIRECCIONES = "direcciones" // 👈 Asegura que exista
        const val TABLE_GUIA = "guia" // 👈 Asegura que exista
        const val TABLE_SOLICITUDES = "solicitudes" // 👈 Asegura que exista
    }

    // ... (El método createTables sigue siendo exactamente el mismo que la versión anterior)
    private fun createTables(db: SQLiteDatabase) {

        // ==========================================================
        // 1. Tabla de Usuarios (Autenticación Global) - CORREGIDA
        // ==========================================================
        db.execSQL(
            """
            CREATE TABLE $TABLE_USERS (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT,
                email TEXT UNIQUE NOT NULL,
                password_hash TEXT NOT NULL,
                phone_number TEXT,
                role TEXT NOT NULL CHECK(role IN ('CLIENTE', 'ADMIN', 'CONDUCTOR', 'GESTOR', 'FUNCIONARIO', 'ANALISTA')),
                sucursal TEXT DEFAULT 'N/A' 
            )
        """
        )

        // ==========================================================
        // 2. Tabla de Recolectores (Información detallada)
        // ==========================================================
        db.execSQL(
            """
            CREATE TABLE $TABLE_RECOLECTORES (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                user_id INTEGER UNIQUE NOT NULL, 
                is_active INTEGER DEFAULT 1,
                FOREIGN KEY(user_id) REFERENCES $TABLE_USERS(id)
            )
        """
        )

        // ==========================================================
        // 3. Tabla de Direcciones
        // ==========================================================
        db.execSQL(
            """
            CREATE TABLE ${DBHelper.TABLE_DIRECCIONES} (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                user_id INTEGER,
                ciudad TEXT NOT NULL,
                full_address TEXT NOT NULL,
                latitude REAL,
                longitude REAL,
                created_at TEXT NOT NULL, -- Usar TEXT para guardar fecha/hora en formato ISO8601
                FOREIGN KEY(user_id) REFERENCES $TABLE_USERS(id)
            )
        """
        )

        // ==========================================================
        // 4. Tabla de Guía / Tracking
        // ==========================================================
        db.execSQL(
            """
            CREATE TABLE guia (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                tracking_number TEXT UNIQUE NOT NULL,
                descripcion TEXT,
                valor REAL DEFAULT 0.0,
                peso REAL DEFAULT 0.0,
                created_at INTEGER NOT NULL
            )
        """
        )

        // ==========================================================
        // 5. Tabla de Solicitudes (Tabla central)
        // ==========================================================
        db.execSQL(
            """
            CREATE TABLE solicitudes (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                user_id INTEGER NOT NULL,
                recolector_id INTEGER, 
                direccion_id INTEGER NOT NULL,
                fecha TEXT NOT NULL,
                franja TEXT NOT NULL,
                notas TEXT,
                zona TEXT NOT NULL,
                guia_id INTEGER,
                estado TEXT NOT NULL DEFAULT 'PENDIENTE',
                confirmation_code TEXT,
                created_at INTEGER NOT NULL,
                FOREIGN KEY(user_id) REFERENCES $TABLE_USERS(id),
                FOREIGN KEY(recolector_id) REFERENCES $TABLE_RECOLECTORES(id), -- ¡CORRECCIÓN AQUÍ!
                FOREIGN KEY(direccion_id) REFERENCES direcciones(id),
                FOREIGN KEY(guia_id) REFERENCES guia(id)
            )
        """
        )

        // ÍNDICES para optimizar las búsquedas
        db.execSQL("CREATE INDEX idx_sol_recolector ON solicitudes(recolector_id)")
        db.execSQL("CREATE INDEX idx_guia_tracking ON guia(tracking_number)")
    }


    /**
     * Inserta un usuario administrador de prueba en la tabla principal de usuarios ($TABLE_USERS).
     * @param db Instancia de la base de datos abierta.
     */
    private fun insertDefaultAdmin(db: SQLiteDatabase) {
        val adminPassword = "Admin123"

        // 🏆 USO DE LA NUEVA CLASE PasswordHasher
        val hashedPassword = PasswordHasher.hashPassword(adminPassword)

        if (hashedPassword.isEmpty()) {
            Log.e("DBHelper", "FATAL: No se pudo generar el hash del password de administrador.")
            return
        }

        val cv = ContentValues().apply {
            put("name", "Administrador Principal")
            put("email", "admin@logistica.com")
            put("password_hash", hashedPassword)
            put("phone_number", "3223691238")
            put("role", "ADMIN")
            put("sucursal", "CENTRAL")
        }
        try {
            db.insert(TABLE_USERS, null, cv)
            Log.d("DBHelper", "Admin insertado correctamente en la tabla $TABLE_USERS.")
        } catch (e: Exception) {
            Log.e("DBHelper", "Error al insertar admin: ${e.message}")
        }
    }

    override fun onCreate(db: SQLiteDatabase) {
        createTables(db)
        insertDefaultAdmin(db)
    }

    /**
     * Se llama cuando la base de datos necesita ser actualizada.
     */
    override fun onUpgrade(db: SQLiteDatabase, oldV: Int, newV: Int) {
        if (oldV < newV) {
            Log.w(
                "DBHelper",
                "Actualizando base de datos de V$oldV a V$newV. ¡Se borrarán todos los datos!"
            )
            // Eliminar todas las tablas (migración destructiva)
            db.execSQL("DROP TABLE IF EXISTS solicitudes")
            db.execSQL("DROP TABLE IF EXISTS direcciones")
            db.execSQL("DROP TABLE IF EXISTS guia")
            db.execSQL("DROP TABLE IF EXISTS $TABLE_RECOLECTORES")
            db.execSQL("DROP TABLE IF EXISTS $TABLE_USERS")

            onCreate(db)
        }
    }
}